### PR TITLE
[DOM-20310] Fix `model_version_publish`

### PR DESCRIPTION
**Pull request by abhijeet2096**
--------------------------------------------------------------------------------------------------------
# Fixed `model_version_publish` 
This method was initially failing because of missing param `projectId` with following error message

## Error Message

```
{'errors': [{'message': 'This field is required', 'param': 'projectId'}]}
```

## Problem

Earlier, the param projectId was found using hitting a GET request on following URI `https://{deployment_url}/u/{owner_name}/{project_name}/endpoints/modelManager` and retrieving `projectId` using Regex. 

 ## Solution

In this P.R I have replaced the routine to find `projectId` with this API
```
GET https://{deployment_url}//v4/gateway/projects/findProjectByOwnerAndName?ownerName={owner_name}&projectName={project_name}
```

##  Verification

Verified this on a deployment with domino version `4.1.9` and `4.2.0-SNAPSHOT`

## Additional changes

* Added check for version compatibility in `Domino` class
* Added `version.py` as single source for current version
* Added `Version-Compatibility` table in `Readme.md`

## Links

[DOM-20310](https://dominodatalab.atlassian.net/browse/DOM-20310)

### CODE REVIEW

File: domino/helpers.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from distutils.version import LooseVersion as parse_version
+from .constants import MINIMUM_SUPPORTED_DOMINO_VERSION
+from version import PYTHON_DOMINO_VERSION
+
+
+def is_version_compatible(version: str) -> None:
```
Comment (dmitriyvolk) [inline domino/helpers.py:6]: I would call this `enforce_version_compatibility` or `ensure_version_compatible` or something similar. In Java, if a function is called `isSomething`, it is assumed to be side-effect free and returning a Boolean. I know Python is not Java, but still...

File: README.md [hunk 2]
```diff
@@ -4,6 +4,15 @@ Python bindings for the Domino API.
 
 Permits interaction with a Domino deployment from Python using the [Domino API](https://dominodatalab.github.io/api-docs/).
 
+## Version Compatibility Matrix
+
+`python-domino` is compatible to domino version as follows:
```
Comment (dmitriyvolk) [inline README.md:9]: Something's off about this sentence. `Domino` should be capitalized, at least. Here's how a similar thing is said in `play-pac4j`:

`Several versions of the library are available for the different versions of the Play framework:`

File: domino/domino.py [hunk 3]
```diff
@@ -41,12 +36,18 @@ def __init__(self, project, api_key=None, host=None):
                 constructor value or through DOMINO_USER_API_KEY environment \
                 variable.")
 
+        self._logger.info('Initializing Domino API with host ' + host)
```
Comment (gsvijayraajaa) [inline domino/domino.py:39]: Try to use f-string's

File: domino/domino.py [hunk 4]
```diff
@@ -41,12 +36,18 @@ def __init__(self, project, api_key=None, host=None):
...
+
+        owner_username, project_name = project.split("/")
+        self._routes = _Routes(host, owner_username, project_name)
+
         # Get version
         self._version = self.deployment_version().get("version")
         print("Your Domino deployment is running \
```
Comment (gsvijayraajaa) [inline domino/domino.py:46]: logger here, no prints !

File: domino/domino.py [hunk 5]
```diff
@@ -380,17 +381,14 @@ def requires_at_least(self, at_least_version):
...
-        response = requests.get(url, auth=('', self._api_key))
-        regex = re.compile("/models/create\\?projectId=(.{24,24})")
-        matches = regex.findall(response.text)
-        if len(matches) > 0:
-            return matches[0]
+        url = self._routes.find_project_by_owner_name_and_project_name()
+        key = "id"
+        response = self._get(url)
+        if key in response.keys():
+            project_id = response[key]
```
Comment (gsvijayraajaa) [inline domino/domino.py:388]:
```
if key in response.keys():
        return(response[key])
```

File: domino/routes.py [hunk 6]
```diff
@@ -105,3 +99,8 @@ def deployment_version(self):
     # App URLs
     def app_publish(self):
         return self._build_project_url_private_api() + '/nb/startSession'
+
+    # Find Project By OwnerName and project name
+    def find_project_by_owner_name_and_project_name(self):
+        return self.host + '/v4/gateway/projects/findProjectByOwnerAndName' + \
```
Comment (gsvijayraajaa) [inline domino/routes.py:105]: combine it as a single f string

File: domino/helpers.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    """
+    Helper function to check for version compatibility
+
+    @:param version  Domino version to check version compatibility against
+
+    @:exception throws exception if domino version isn't compatible with
+                current python-domino version
+    """
+    current_version = parse_version(version)
```
Comment (gsvijayraajaa) [inline domino/helpers.py:15]: ✂️ this var isnt used in more than one place

File: domino/helpers.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    minimum_supported_version = parse_version(MINIMUM_SUPPORTED_DOMINO_VERSION)
```
Comment (gsvijayraajaa) [inline domino/helpers.py:16]: same here

File: domino/routes.py [hunk 9]
```diff
@@ -105,3 +99,8 @@ def deployment_version(self):
     # App URLs
     def app_publish(self):
         return self._build_project_url_private_api() + '/nb/startSession'
+
+    # Find Project By OwnerName and project name
+    def find_project_by_owner_name_and_project_name(self):
```
Comment (gsvijayraajaa) [inline domino/routes.py:104]: The function name is misleading, this is just building the url as opposed to actually returning the actual value.

File: README.md [hunk 10]
```diff
@@ -4,6 +4,15 @@ Python bindings for the Domino API.
 
 Permits interaction with a Domino deployment from Python using the [Domino API](https://dominodatalab.github.io/api-docs/).
 
+## Version Compatibility Matrix
+
+Several versions of the `python-domino` library are available for the different versions of the `Domino`:
```
Comment (gsvijayraajaa) [inline README.md:9]: The gist should just be something like the compatibility between domino version and the python-domino library.

File: domino/helpers.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
+from distutils.version import LooseVersion as parse_version
+from .constants import MINIMUM_SUPPORTED_DOMINO_VERSION
+from version import PYTHON_DOMINO_VERSION
+
+
+def enforce_version_compatibility(version: str) -> None:
```
Comment (gsvijayraajaa) [inline domino/helpers.py:6]: Do not like the `enforce` part. We are just checking the version compatability

File: domino/domino.py [hunk 12]
```diff
@@ -41,12 +37,21 @@ def __init__(self, project, api_key=None, host=None):
                 constructor value or through DOMINO_USER_API_KEY environment \
                 variable.")
 
+        self._logger.info(f'Initializing Domino API with host {host}')
+
+        owner_username, project_name = project.split("/")
+        self._routes = _Routes(host, owner_username, project_name)
+
         # Get version
```
Comment (gsvijayraajaa) [inline domino/domino.py:45]: ✂️ understandable from the code

File: domino/domino.py [hunk 13]
```diff
@@ -41,12 +37,21 @@ def __init__(self, project, api_key=None, host=None):
...
         self._version = self.deployment_version().get("version")
-        print("Your Domino deployment is running \
-              version {}".format(self._version))
+        self._logger.info(f"Domino deployment {host} is running version {self._version}")
 
-        # set project ID to blank, we may or may not need it later
+        # Check version compatibility
```
Comment (gsvijayraajaa) [inline domino/domino.py:49]: ✂️ understandable from the code

File: domino/routes.py [hunk 14]
```diff
@@ -88,12 +88,6 @@ def model_versions_get(self, model_id):
     def model_version_publish(self, model_id):
         return self._build_models_url() + '/' + model_id + '/versions'
```
Comment (gsvijayraajaa) [inline domino/routes.py:89]: Can you try to have fstring everywhere for string building? I am not sure, if you would want to clean up existing code.

File: domino/routes.py [hunk 15]
```diff
@@ -105,3 +99,8 @@ def deployment_version(self):
     # App URLs
     def app_publish(self):
         return self._build_project_url_private_api() + '/nb/startSession'
+
+    # Find Project By OwnerName and project name Url
```
Comment (gsvijayraajaa) [inline domino/routes.py:103]: In general would be a good practise to have the comments as doc string.

File: README.md [hunk 16]
```diff
@@ -4,6 +4,15 @@ Python bindings for the Domino API.
...
 Permits interaction with a Domino deployment from Python using the [Domino API](https://dominodatalab.github.io/api-docs/).
 
+## Version Compatibility Matrix
+
+`python-domino` library is compatible with different versions of the `Domino`:
+
+| Domino Versions | Python-Domino |
+| --------------- |:-------------:|
+| 3.6.x or Lower  | 0.3.5         |
+| 4.0.x or Higher | 1.0.0         |
```
Comment (ddl-alexpanin) [inline README.md:14]: We're releasing new version for 4.2, so probably should be:

4.1.x or Lower  0.3.5
4.2 or Higher   1.0.0

File: domino/constants.py [hunk 17]
```diff
@@ -0,0 +1,5 @@
+"""
+Minimum Domino version supported by this python-domino library
+"""
+
+MINIMUM_SUPPORTED_DOMINO_VERSION = '4.0.0'
```
Comment (ddl-alexpanin) [inline domino/constants.py:5]: Should it be '4.2' instead? Not sure if '4.2.0' or what we actually have, `cloud-develop` reports 4.2.0-SNAPSHOT

File: domino/helpers.py [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+def is_version_compatible(version: str) -> bool:
+    """
+    Helper function to check for version compatibility
+
+    @:param version  Domino version to check version compatibility against
+
+    @:return bool   Boolean representing if version is compatible or not
+    """
+    return parse_version(version) > parse_version(MINIMUM_SUPPORTED_DOMINO_VERSION)
```
Comment (ddl-alexpanin) [inline domino/helpers.py:13]: make sure "greater" works if minimum is 4.2. Maybe greater or equal?

**Pull request merged**